### PR TITLE
Move log timestamp to Foundation Date

### PR DESCRIPTION
**Pull request by JGiola**
# Summary

This pull request is aimed to the issue #129 for using Foundation.Date for log timestamps
# Design decisions

I've add a private time formatter shared instance for keeping the initialisation cost low, and set the format template to print hours, minutes, seconds and milliseconds.
I don't know if you prefer additional information on the date or the time is enough.

### CODE REVIEW

**Comment by dreymonde** [comment]
Nice job, thanks! However, I think it is more appropriate to set date format like this:

``` swift
formatter.dateFormat = DateFormatter.dateFormat(fromTemplate: "HH:mm:ss.SSS", options: 0, locale: .current)
```

This way, the user's locale will be respected. For example, for me it prints time like this:

> 13:52:19,075

(with comma after 19, not the dot), which is appropriate in my region (Ukraine).

**Review by dreymonde — COMMENTED** [review]
Also, maybe it's better to put `dateFormatter` inside `Logger` as `fileprivate` property.

**Comment by JGiola** [comment]
Oh, ok, I've done that at first but then I've gone to do a fixed format.

The dateFormatter property is outside Logger for having it shared across all Logger instances and not create a new one every time a Logger object is created.

I will push the changes when I will be back home.

**Comment by JGiola** [comment]
While I'm at it, you think I have to change the `HH` part of the template with `jj` to adapt dinamically to 12/24 hour clock?
Or at least is what I can understand from this [documentation](http://www.unicode.org/reports/tr35/tr35-31/tr35-dates.html#Date_Format_Patterns) even if I don't know if you have to add the `a` to add AM/PM...

File: Modules/Axis/Sources/Axis/Log/Log.swift [hunk 1]
```diff
@@ -51,6 +47,13 @@ public final class Logger {
         self.appenders.append(contentsOf: appenders)
         self.name = name
     }
+
+    fileprivate static let logTimeFormatter: DateFormatter = {
+        let formatter = DateFormatter()
+        let format = DateFormatter.dateFormat(fromTemplate: "HH:mm:ss.SSS", options: 0, locale: .current)
```
Comment (Danappelxx) [inline Modules/Axis/Sources/Axis/Log/Log.swift]: Could we also include the day/month/year and locale here? I think that would be useful. Ideal date IMO: `10/06/16 12:42:10.123 GMT-8`.

Note - I'm not sure if `DateFormatter` supports figuring out whether the month should go before or in-between the day and year (in the US, we have the day in the middle).

I also think we should use the 24-hour system regardless of the locale, so the code you have now (`HH`) is good.

**Review by Danappelxx — CHANGES_REQUESTED** [review]
Great PR! Just need to figure out the format and it's good to go.

**Comment by JGiola** [comment]
No problem, for the mont and day I'm sure it can do if you use the enum specifiers, I don't know if it will do the same with the localised template...

**Comment by JGiola** [comment]
Done, I've even removed the separator from the template because they aren't needed, foundation is figuring out everything by itself.

And yes, the order of year, month and day are based on the locale 👍🏻

**Comment by Danappelxx** [comment]
Awesome! Just one more thing - could you write a unit test that ensures the formatter spits out the right string? I believe what you say about the unnecessary separators but I'm worried that it might be different on Linux for whatever reason, etc.

**Review by Danappelxx — APPROVED** [review]

**Comment by JGiola** [comment]
The tests are no problem, but it possible to cheat a bit? 😁
Because the actual lotTimeFormatter is a fileprivate property I can't access directly inside the tests and in addition to that I have to set a fixed sets of locale and time zone to know how the strings would come out and so reset the dateFormat to it.

My idea is to tests on a brand new DateFormatter with a couple of different locale and time zone and two or three different dates to see if the format string without any separator is always equals to something we will know.
Is it ok for you?

**Comment by Danappelxx** [comment]
If you use `@testable import` you should be able to access the `fileprivate` property. If this does not work, maybe make the formatter public? That could provide other benefits (user customization) anyway. 

I agree that you should use several constant dates with locales and just plug them into the formatter.

**Comment by JGiola** [comment]
`@testable import`allow only internal properties to be accessible. I've moved the `logTimeFormatter` property to be public so people can changes all the properties as they like.

The tests will check against the en_US and it_IT locales, I have to add another one?

**Comment by JGiola** [comment]
I'm start to thinking that on Linux the unicode library is parsing the date skeleton template in a totally different way than on macOS.
It seems to completely ignore the forced 24 hour marker and for the Italian locale the two platform don't agree on the coma separator between date and hour...

**Comment by JGiola** [comment]
Ok, I've spin up an ubuntu 12.04 server and I've found a super weird bug. On Linux if you add anyone of the time zone specifier to the skeleton format will disable the forced 24h format, as soon as you change the skeleton from `yyMMdd HHmmssSSSz` to `yyMMdd HHmmssSSS` the printed hours will revert correctly to the force 24h format.

**Comment by Danappelxx** [comment]
Interesting... so what you're saying is that removing the z will fix it? Can you try committing that change? If it fixes CI, we can happily merge :)

**Comment by Danappelxx** [comment]
Oh, so removing Z removes the time zone? That's a shame... @paulofaria @robertjpayne what do you think is the right thing to do here?

**Comment by paulofaria** [comment]
@JGiola could you please raise a bug at bug.swift.org? I think the formatting itself is secondary in regards to correctness. Timezones should be respected above all. If this means different formatting, I'm fine with that.

**Comment by JGiola** [comment]
Yeah, I've filed it when I've been able to check what was going on on a Ubuntu server.
If you want to join in and write it down better is here https://bugs.swift.org/browse/SR-2906

**Comment by JGiola** [comment]
In regard of the pull request what you want me to do? Going back on adding the time zone to the skeleton format and comment out the tests for now? Or leave things as they are now?

**Pull request merged**